sway-core/src/semantic_analysis/ast_node/expression/match_expression/typed/matcher.rs

```diff
@@ -1,38 +1,142 @@
+use either::Either;
 use indexmap::IndexMap;
 
 use crate::{
-    language::{ty, CallPath, Literal},
+    language::{
+        ty::{self, TyConstantDecl},
+        CallPath, Literal,
+    },
     semantic_analysis::{
         ast_node::expression::typed_expression::{
-            instantiate_struct_field_access, instantiate_tuple_index_access,
-            instantiate_unsafe_downcast,
+            instantiate_enum_unsafe_downcast, instantiate_struct_field_access,
+            instantiate_tuple_index_access,
         },
         TypeCheckContext,
     },
-    Ident, TypeId, UnifyCheck,
+    Ident, TypeId, TypeInfo, UnifyCheck,
 };
 
 use sway_error::{
     error::CompileError,
     handler::{ErrorEmitted, Handler},
 };
 
-use sway_types::{span::Span, Spanned};
+use sway_types::{integer_bits::IntegerBits, span::Span, Named, Spanned};
+
+/// A single requirement in the form `<lhs> == <rhs>` that has to be
+/// fulfilled for the match arm to match.
+pub(super) type MatchReq = (ty::TyExpression, ty::TyExpression);
+
+/// A single variable in the form `let <ident> = <expression>`
+/// that has to be extracted from the match arm.
+pub(super) type MatchVarDecl = (Ident, ty::TyExpression);
+
+/// A leaf of a match pattern can be either a requirement on the scrutinee or a
+/// variable declaration but not both at the same time.
+/// In the case of the catch-all `_` we will have neither a requirement nor
+/// a variable declaration.
+pub(super) type ReqOrVarDecl = Option<Either<MatchReq, MatchVarDecl>>;
```

Comment:
Instead of using `Option<Either<...>>` for a ternary union I think it would be clearer to use an enum with proper names for the cases that you later describe for this.

Reply:
Agree! Done!